### PR TITLE
Refactor ops

### PR DESCRIPTION
**Pull request by niboshi**
"ops" and "routines" are one of the difficult parts of the core code, because these ad-hoc data structures are not documented.

This PR makes ops classes (`_Ops` and `_Op`) and moves related functions into them, in the hope of making it easier to understand.

### CODE REVIEW

**Comment by asi1024** [comment]
Could you fix for failure of `tests/cupy_tests/math_tests`?

File: cupy/core/_kernel.pyx [hunk 1]
```diff
@@ -1008,10 +961,105 @@ cpdef create_ufunc(name, ops, routine=None, preamble='', doc='',
...
+        self.ops = ops
+        self.nin = nin
+        self.nout = nout
+
+    @staticmethod
+    cdef _Ops from_tuples(object ops, routine):
+        ops_ = []
+        for t in ops:
+            if isinstance(t, tuple):
+                typ, rt = t
```
Comment (asi1024) [inline cupy/core/_kernel.pyx]: Some elements in `rt` can be `None`. Please add preprocess like as `rt = tuple([i or j for i, j in zip(rt, routine)])`.

Comment (asi1024) [inline cupy/core/_kernel.pyx]: Some elements in tuple `rt` can be `None`. The following example fails with compilation error.
```py
x = cupy.arange(10, dtype='float16')
x.sum()
```

Could you fix as follows?

```suggestion
                typ, rt = t
                if isinstance(rt, tuple):
                    rt = tuple([i or j for i, j in zip(rt, routine)])
```

**Comment by asi1024** [comment]
@niboshi Could you resolve conflicts?

**Comment by niboshi** [comment]
PTAL

**Comment by niboshi** [comment]
LGTM

**Comment by niboshi** [comment]
PTAL

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**